scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala

```diff
@@ -79,100 +80,91 @@ private[scio] class KryoAtomicCoder[T](private val options: KryoOptions) extends
 
   import KryoAtomicCoder.logger
 
-  private val kryoState: ThreadLocal[KryoState] =
-    new EmptyOnDeserializationThreadLocal[KryoState] {
-      override def initialValue(): KryoState = {
-        val k = KryoSerializer.registered.newKryo()
-
-        k.setReferences(options.referenceTracking)
-        k.setRegistrationRequired(options.registrationRequired)
-
-        k.forClass(new CoderSerializer(InstantCoder.of()))
-        k.forClass(new CoderSerializer(TableRowJsonCoder.of()))
-
-        // Java Iterable/Collection are missing proper equality check, use custom CBF as a
-        // workaround
-        k.register(
-          classOf[Wrappers.JIterableWrapper[_]],
-          new JTraversableSerializer[Any, Iterable[Any]]()(new JIterableWrapperCBF[Any]))
-        k.register(
-          classOf[Wrappers.JCollectionWrapper[_]],
-          new JTraversableSerializer[Any, Iterable[Any]]()(new JCollectionWrapperCBF[Any]))
-        // Wrapped Java collections may have immutable implementations, i.e. Guava, treat them as
-        // regular Scala collections as a workaround
-        k.register(
-          classOf[Wrappers.JListWrapper[_]], new JTraversableSerializer[Any, mutable.Buffer[Any]])
-
-        k.forSubclass[SpecificRecordBase](new SpecificAvroSerializer)
-        k.forSubclass[GenericRecord](new GenericAvroSerializer)
-        k.forSubclass[Message](new ProtobufSerializer)
-
-        k.forSubclass[LocalDate](new JodaLocalDateSerializer)
-        k.forSubclass[LocalTime](new JodaLocalTimeSerializer)
-        k.forSubclass[LocalDateTime](new JodaLocalDateTimeSerializer)
-        k.forSubclass[DateTime](new JodaDateTimeSerializer)
-
-        k.forSubclass[Path](new JPathSerializer)
-        k.forSubclass[ByteString](new ByteStringSerializer)
-
-        k.forClass(new KVSerializer)
-        // TODO:
-        // TimestampedValueCoder
-
-        new AlgebirdRegistrar()(k)
-        KryoRegistrarLoader.load(k)
-
-        val input = new InputChunked(options.bufferSize)
-        val output = new OutputChunked(options.bufferSize)
-
-        KryoState(k, input, output)
-      }
+  private val kryoPool: EmptyOnSerializeAtomicReference[ObjectPool[KryoState]]
+  = new EmptyOnSerializeAtomicReference[ObjectPool[KryoState]]() {
```

Comment:
Why do you need that intermediate `EmptyOnSerializeAtomicReference` here. I don't get the need for an `AtomicReference``